keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,632 @@
+  // min quota is unmet.
+  preemption ClusterQueuePreemption
+}
+
+type ClusterQueuePreemption struct {
+  // waitBeforePreemptionSeconds is the time in seconds that the ClusterQueue
+  // is willing to wait for Workloads to finish since the first time that
+  // the Workload was attempted for admission, before preempting active
+  // Workloads with lower priority.
+  // The time when the Workload was first attempted for admission is present
+  // in the transitionTimestamp of the Admitted condition.
+  // If null, it means to wait indefinitely, effectively disabling preemption
+  // for incoming Workloads in this ClusterQueue.
+  WaitBeforePreemptionSeconds *int64
+}
+```
+
+### Changes in scheduling algorithm
+
+The following changes in the scheduling algorithm are required to implement
+preemption.
+
+#### Detecting Workloads that might benefit from preemption
+
+The first stage during scheduling is to assign flavors to each resource of
+a workload.
+
+The algorithm is like follows:
+
+    For each resource (or set of resources with the same flavors), evaluate
+    flavors in the order established in the ClusterQueue*:
+
+    0. Find a flavor that still has quota in the cohort (borrowing allowed),
+       but doesn't surpass the max quota for the CQ. Keep track of whether
+       borrowing was needed.
+    1. [New step] if no flavor was found, find a flavor that is under min quota
+       for the CQ, regardless of the cohort usage.
+    2. [New step] if no flavor was found, use the first flavor in the list that
+       has more quota than the Workload request.
+
+
+We will call the flavor assignments type 0, 1 or 2, depending on which step they
+were found.
+    
+A flavor assignment of type 1 means that we need to preempt or wait for other
+workloads in the cohort to finish to accommodate this workload, because the
+ClusterQueue is lending its resources.
+
+A flavor assigment of type 2 means that we need to preempt or wait for other
+workloads in the ClusterQueue to finish to accomodate this workload.
+
+Note that a Workload could get assignments of all types.
+
+[#312](https://github.com/kubernetes-sigs/kueue/issues/312) discusses different
+strategies to select a flavor.
+
+#### Sorting Workloads that are heads of ClusterQueues
+
+Sorting uses the following criteria:
+
+1. Flavors that don't borrow first.
+2. [New criteria] Highest priority first.
+3. Older creation timestamp first.
+
+Note that these criteria might put Workloads that require preemption ahead,
+because preemption doesn't require borrowing more resources. This is desired,
+because preemption to recover quota or admit high priority Workloads takes
+preference over borrowing.
+
+#### Admission
+
+When iterating over workloads to be admitted, in the order given by the previous
+section, we disallow borrowing in the cohort in the current cycle after
+evaluating a Workload that doesn't require borrowing (including Workloads that
```

Comment:
ok, so this is the same as what we do now (i.e., a workload was admitted to the cohort), plus (and this is the difference from the current logic) workloads that preempt others, correct?